external/fv3fit/fv3fit/pytorch/cyclegan/train.py

```diff
@@ -0,0 +1,145 @@
+from fv3fit._shared.hyperparameters import Hyperparameters
+import dataclasses
+import tensorflow as tf
+from fv3fit.pytorch.predict import PytorchPredictor
+from fv3fit.pytorch.loss import LossConfig
+from fv3fit.pytorch.optimizer import OptimizerConfig
+from fv3fit.pytorch.training_loop import TrainingConfig
+
+from fv3fit._shared import register_training_function
+from typing import (
+    List,
+    Optional,
+    Tuple,
+)
+from fv3fit.tfdataset import ensure_nd, apply_to_mapping
+from .network import Generator
+from fv3fit.pytorch.graph.train import (
+    get_scalers,
+    get_mapping_scale_func,
+    get_Xy_dataset,
+)
+from toolz import curry
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class GeneratorConfig:
+    n_convolutions: int = 3
+    n_resnet: int = 3
+    max_filters: int = 256
+
+
+@dataclasses.dataclass
+class AutoencoderHyperparameters(Hyperparameters):
+
+    state_variables: List[str]
+    normalization_fit_samples: int = 50_000
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("AdamW")
+    )
+    generator: GeneratorConfig = dataclasses.field(
+        default_factory=lambda: GeneratorConfig()
+    )
+    training_loop: "TrainingConfig" = dataclasses.field(
+        default_factory=lambda: TrainingConfig()
+    )
+    loss: LossConfig = LossConfig(loss_type="mse")
+    noise_amount: float = 0.5
+
+    @property
+    def variables(self):
+        return tuple(self.state_variables)
+
+
+@curry
+def define_noisy_input(data: tf.Tensor, stdev=0.1) -> Tuple[tf.Tensor, tf.Tensor]:
```

Comment:
Is the noisy input really a feature of the autoencoder, or is this more part of the test framework you've chosen? Feels like the latter, in which case the code related to it should move there.

Reply:
OK, that's fine then.